File: dotnet/src/Connectors/Connectors.UnitTests/Memory/AzureCognitiveSearch/AzureCognitiveSearchMemoryTests.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Net.Http;
using System.Threading.Tasks;
using Azure.Core;
using Microsoft.SemanticKernel.Connectors.Memory.AzureCognitiveSearch;
using Xunit;

namespace SemanticKernel.Connectors.UnitTests.Memory.AzureCognitiveSearch;

/// <summary>
/// Unit tests for <see cref="AzureCognitiveSearchMemory"/> class.
/// </summary>
public sealed class AzureCognitiveSearchMemoryTests : IDisposable
{
    private HttpMessageHandlerStub messageHandlerStub;
    private HttpClient httpClient;

    public AzureCognitiveSearchMemoryTests()
    {
        this.messageHandlerStub = new HttpMessageHandlerStub();

        this.httpClient = new HttpClient(this.messageHandlerStub, false);
    }

    [Fact]
    public async Task CustomHttpClientProvidedToFirstConstructorShouldBeUsed()
    {
        //Arrange
        var sut = new AzureCognitiveSearchMemory("https://fake-random-test-host/fake-path", "fake-api-key", this.httpClient);

        //Act
        await sut.GetAsync("fake-collection", "fake-query");

        //Assert
        Assert.StartsWith("https://fake-random-test-host/fake-path/indexes('fake-collection')", this.messageHandlerStub.RequestUri?.AbsoluteUri, StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public async Task CustomHttpClientProvidedToSecondConstructorShouldBeUsed()
    {
        //Arrange
        var credentials = DelegatedTokenCredential.Create((_, __) => new AccessToken("fake-access-token", DateTimeOffset.UtcNow.AddMinutes(15)));

        var sut = new AzureCognitiveSearchMemory("https://fake-random-test-host/fake-path", credentials, this.httpClient);

        //Act
        await sut.GetAsync("fake-collection", "fake-key");

        //Assert
        Assert.StartsWith("https://fake-random-test-host/fake-path/indexes('fake-collection')", this.messageHandlerStub.RequestUri?.AbsoluteUri, StringComparison.OrdinalIgnoreCase);
    }

    public void Dispose()
    {
        this.httpClient.Dispose();
        this.messageHandlerStub.Dispose();
    }
}

File: java/semantickernel-connectors-parent/semantickernel-connectors/src/test/java/com/microsoft/semantickernel/connectors/memory/azurecognitivesearch/AzureCognitiveSearchMemoryTests.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.connectors.memory.azurecognitivesearch;

import com.azure.core.credential.AzureKeyCredential;
import com.azure.core.http.HttpClient;
import com.azure.core.http.HttpRequest;
import com.azure.core.http.HttpResponse;
import com.azure.search.documents.indexes.SearchIndexAsyncClient;
import com.azure.search.documents.indexes.SearchIndexClientBuilder;
import com.microsoft.semantickernel.Kernel;
import com.microsoft.semantickernel.builders.SKBuilders;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class AzureCognitiveSearchMemoryTests
{

    SearchIndexAsyncClient searchIndexAsyncClient;
    HttpResponse httpResponse;

    @BeforeEach
    void setUp() {
        httpResponse = Mockito.mock(HttpResponse.class);

        HttpClient httpClient = Mockito.mock(HttpClient.class);
        Mockito.when(httpClient.send(Mockito.any(HttpRequest.class))
                .thenReturn(Mono.just(httpResponse)));

        searchIndexAsyncClient = new SearchIndexClientBuilder()
                .endpoint("https://fake-random-test-host/fake-path")
                .credential(new AzureKeyCredential("fake-key"))
                .httpClient(httpClient)
                .buildAsyncClient();
    }

    @Test @Disabled
    void AzureCognitiveSearchMemoryStoreShouldBeProperlyInitialized() {
        //Arrange
        Mockito.when(httpResponse.getBody())
                .thenReturn(Flux.just(
                        ByteBuffer.wrap(
                                "{\"value\": [{\"name\": \"fake-index1\"}]}"
                                        .getBytes(StandardCharsets.UTF_8))));
        Mockito.when(httpResponse.getStatusCode()).thenReturn(200);

        Kernel kernel = SKBuilders.kernel()
            .withMemory(new AzureCognitiveSearchMemory(searchIndexAsyncClient))
            .build();
    //Act
    //This call triggers a subsequent call to Azure Cognitive Search Memory store.
    List<String> collections = kernel.getMemory().getCollectionsAsync().block();

    //Assert
//    assertEquals("https://fake-random-test-host/fake-path/indexes?$select=%2A&api-version=2021-04-30-Preview"), this.messageHandlerStub?.RequestUri?.AbsoluteUri);

//    var headerValues = Enumerable.Empty<string>();
//    var headerExists = this.messageHandlerStub?.RequestHeaders?.TryGetValues("Api-Key", out headerValues);
//    Assert.True(headerExists);
    assertTrue(collections.contains("fake-index1"));
}

}
